smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java

```diff
@@ -0,0 +1,242 @@
+
+                if (blockListCached == null) {
+                    blockListCached = new ArrayList<Jid>();
+                }
+
+                Jid unblockedJid = unblockContactIQ.getJid();
+                if (unblockedJid == null) { // remove all
+                    blockListCached.clear();
+                } else { // remove only one
+                    removeFromBlockList(unblockedJid);
+                }
+
+                return IQ.createResultIQ(unblockContactIQ);
+            }
+        });
+
+        connection.addConnectionListener(new AbstractConnectionListener() {
+            @Override
+            public void authenticated(XMPPConnection connection, boolean resumed) {
+                // No need to reset the cache if the connection got resumed.
+                if (resumed) {
+                    return;
+                }
+                blockListCached = null;
+            }
+        });
+    }
+
+    private void addToBlockList(Jid blockedJid) {
+        if (searchJid(blockedJid) == -1) {
+            blockListCached.add(blockedJid);
+        }
+    }
+
+    private void removeFromBlockList(Jid unblockedJid) {
+        int position = searchJid(unblockedJid);
+        if (position != -1) {
+            blockListCached.remove(position);
+        }
+    }
+
+    private int searchJid(Jid jid) {
+        int i = -1;
+        for (int j = 0; j < blockListCached.size(); j++) {
+            if (blockListCached.get(j).equals(jid)) {
+                i = j;
+            }
+        }
+        return i;
+    }
+
+    /**
+     * Returns true if Blocking Command is supported by the server.
+     * 
+     * @return true if Blocking Command is supported by the server.
+     * @throws NoResponseException
+     * @throws XMPPErrorException
+     * @throws NotConnectedException
+     * @throws InterruptedException
+     */
+    public boolean isSupportedByServer()
+            throws NoResponseException, XMPPErrorException, NotConnectedException, InterruptedException {
+        return ServiceDiscoveryManager.getInstanceFor(connection()).serverSupportsFeature(NAMESPACE);
+    }
+
+    /**
+     * Returns the block list.
+     * 
+     * @return the blocking list
+     * @throws NoResponseException
+     * @throws XMPPErrorException
+     * @throws NotConnectedException
+     * @throws InterruptedException
+     */
+    public List<Jid> getBlockList()
+            throws NoResponseException, XMPPErrorException, NotConnectedException, InterruptedException {
+
+        if (blockListCached != null) {
+            return blockListCached;
+        }
+
+        BlockListIQ blockListIQ = new BlockListIQ();
+        BlockListIQ blockListIQResult = connection().createPacketCollectorAndSend(blockListIQ).nextResultOrThrow();
+        blockListCached = blockListIQResult.getJids();
+        return (blockListCached == null) ? null : Collections.unmodifiableList(blockListCached);
```

Comment:
Methods like these should never return `null` when they could return just an empty list. This removes the burden from the caller to handle the empty list case by checking for `null`.
But this is not only true for the high lever parts of the API, i.e. here, but also for the lower ones: `BlockListIQ.getJids()` should also never return `null`, but `Collections.emptyList()`.